### ISSUE TITLE
Parallel transactions in knex

### ISSUE THREAD

**Issue by nicholaswmin** [tgriesser/knex#1806]
Is there any way to get a transaction working in parallel?

- Knex [doesn't support multiple queries in one go][2]
- A stop-gap solution could be to enable connection pooling and execute the queries in parallel, using [`Promise.all()`][3]

### Here's the problem

A transaction needs to execute all it's queries on the same connection, thus calls to `Promise.all()` are fast when executed *not as part of a transaction*, and slow when executed as part of a transaction

    // can be called as part of a transaction, or not
    getAllData: function(trx, idCustomer) {
      return new Promise((resolve, reject) => {
        let data = {};
        
        Promise.all([
          getFoo(trx); // knex query using `trx` as the knex instance
          getBar(trx); // knex query using `trx` as the knex instance
        ]).then((result) => {
          data.foo = result[0];
          data.bar = result[1];
          resolve(data);
        }).catch((err) => {
          reject(err);
        })
      });
    });

I could use `knex.raw()` and do actual multiple-statements queries in one go but that would couple the DB calls with the DB,thus eliminating knex's added value of being a DB abstraction layer

  [1]: http://knexjs.org/
  [2]: https://github.com/tgriesser/knex/issues/257#issuecomment-41682716
  [3]: https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Promise/all

**Comment by elhigu** [comment]
If you mean by "making transaction parallel", that multiple queries are sent to the same transaction using different connections, it is not possible. It is limitation of the database server itself.

If you are referring that multiple queries are sent to DB as fast as possible without waiting query response for each query before sending the next one... At least postgres driver doesn't seem to support that https://github.com/brianc/node-postgres/issues/896. 

Most probably `Promise.all` doesn't help in this case and just uses more resources / buffers when node is trying to do parallel stuff that will be sent to DB one by one.

**Comment by nicholaswmin** [comment]
> If you are referring that multiple queries are sent to DB as fast as possible without waiting query response for each query before sending the next one

Isn't there any way to handle that on knex-level? 

Not sure if I'm missing something but maybe the `trx` object could accumulate the queries itself and make the decision when to actually initiate the DB call, no?

**Comment by elhigu** [comment]
@nicholaswmin not really, it is not knex responsibility to decide how communication between driver and database is handled. In theory with mysql driver it might be possible to implement (since the driver supports sending multiple queries in one statement), but it would also disable support for sending queries as prepared statements, which is pretty important part of preventing sql-injections.

If your performance requirements are really that high, then you can just use mysql driver directly for those parts of the application.

In db driver level this kind of functionality should be easier to implement. One just would need to send all new queries straight away to DB without waiting for result and add some bookkeeping (maybe just fifo) to be able to associate each result to correct query callback (or on the other hand it might be really tricky, since I haven't looked much into DB driver codes).

**Comment by nicholaswmin** [comment]
@elhigu  Thanks again, I've asked on tedious about this which seems to be the underlying driver of node-mssql which is what I use. Let's see

**Comment by mercmobily** [comment]
This has been answered, closing it. I wonder why it wasn't closed in the first place!
